Add Redistribute transform to model, Java SDK, and most active runners

kennknowles opened the pull request:
This initial PR adds Redistribute transform(s) according to:

 - [Design doc for Redistribute](https://s.apache.org/beam-redistribute) ([dev@ thread](https://lists.apache.org/thread/rc4tcq7k0950cdgw9lyd0s7vkz8rz5s9))
 - [Design doc for Redistribute allowing duplicates](https://s.apache.org/beam-redistribute-allowing-duplicates) ([dev@ thread](https://lists.apache.org/thread/y7zzfc21804lwnwslbhvg6gq7scw2dso))

The various runners were unable to pass the tests without specialized translation, which suggests bugs in their GBK implementation actually. But I just went ahead and implemented specialized translations to match `Reshuffle`.

*Note that the commits in the PR are independent and should not be squashed*

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

kennknowles commented:
Adding R from my collaborators and people who spoke up on dev threads.

R: @Naireen 
R: @lostluck
R: @je-ik

kennknowles commented:
One question which wasn't in the design docs was how to implement: wrap Reshuffle (aka build a composite that just invokes Reshuffle and relies on everything built around it) or fork. This PR chose to fork. Pro/con:

Arguments for wrapping:
 - less code
 - runners that implement Reshuffle specially already will implement Redistribute the same way
 - if there is something I missed in how Reshuffle is treated, it will get picked up because we are still using it

Arguments for forking:
 - decouple whatever state a runner may store, and just generally decouple their evolution
 - people won't unpack their "Redistribute" and see a Reshuffle inside and get the wrong idea
 - (minor) can remove update compatibility path
 - if there is something I missed in how Redistribute and Reshuffle are different, we are free for them to diverge

So I chose forking but could be convinced otherwise. My way creates more code and more work, because we need to make runners treat it specially - for Dataflow we can do way better than the existing Reshuffle translation, for the other runners it'll be a re-use of existing Reshuffle translation.

kennknowles commented:
OK this is ready for review. I don't know if you all are interested in reviewing it or I will also CC @robertwb. I think the separate commits might be easier to review separately (I'm happy to put them in their own PR but also just focusing a review on a particular code root is equivalent)

1. model proto
2. Java SDK impl with translation to proto
3. Each runner - I did Dataflow first but then Spark had test failures so I guess basic GBK fails to preserve metadata so I just went ahead and added translations to Spark, Flink, Samza too.

kennknowles commented:
Redistribute itself and the ValidatesRunner tests are lifted straight from Reshuffle so equivalent levels of assurance FWIW.

kennknowles commented on runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchPortablePipelineTranslator.java:
```diff
@@ -290,6 +296,26 @@ public FlinkPortablePipelineTranslator.Executor translate(
+            Iterables.getOnlyElement(transform.getTransform().getInputsMap().values()));
+    context.addDataSet(
+        Iterables.getOnlyElement(transform.getTransform().getOutputsMap().values()),
+        inputDataSet.rebalance());
```
I think for FlinkRunner today, all of Reshuffle, RedistributeArbitrarily, and RedistributeByKey are basically the same semantics. I do think that is a coincidence rather than a logical necessity, which is why I gave them each their own method.

(these are basically just copy/paste of Reshuffle translation)

kennknowles replied:
Tried a rewrite. still testing it but I pushed the change up.

je-ik replied:
This can be resolved later.

kennknowles commented on sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/RedistributeTest.java:
```diff
@@ -0,0 +1,396 @@
+
+/** Tests for {@link Redistribute}. */
+@RunWith(JUnit4.class)
+public class RedistributeTest implements Serializable {
```
Ah yes, I translated all tests of `Reshuffle.of()` to `Redistribute.arbitrariliy()`. FWIW I think letting the runner control the redistribute is definitely the best technical direction so that is the primary use case. But I change one back so we have some explicit coverage. I could add more.

The invariant for both transforms is simply that the output equals the input.

je-ik replied:
I would suppose, that `Redistribute.byKey()` has additional semantics, or what would be the difference from `.arbitrarily()`? From my understanding the post-redistribute PCollection has data with same key co-located on the same partitions (workers). Is that not the case?

kennknowles replied:
`Redistribute.byKey()` lets the user include keys as a hint for the runner. In both cases, the transform should have output equal to input.

je-ik replied:
If there are no semantic guarantees regarding how the key will be used then it's probably fine. Personally I'd prefer to implement `.arbitrarily()` using `keyBy()` with key as hash of value, but the fact that keys are co-located in partitions cannot be (currently) leveraged, so it probably does not matter.

FYI, I (curiously completely independently) found a bug in Flink's current batch reshuffle: https://github.com/apache/beam/pull/30622.

kennknowles replied:
Yea I agree with what you are saying. Some runners may have even better ways to redistribute arbitrarily (such as using dynamic runtime information to rebalance as it executes). But for Flink now I totally agree with your change, and I will try to mimic or leverage it.

je-ik replied:
I got more information about the original change from `.rebalance()` to the hint. The problem was actually a bug in Flink: https://lists.apache.org/thread/tsbyzlk8p7m59qbgtgfm4rl0cx1rnz1j

I will try to get more info if this has been resolved and if we can use simple `rebalance()`.

je-ik replied:
After offline discussion with @dmvk I think we can try to return back to the hint, but use the `_HASH` variant. According to the discussion this should be respected and have the required semantics (i.e. `Reshuffle` + `Redistribute.arbitrarily()`). I will try it next week.

kennknowles replied:
For this PR I hit some snags in the portable runner in every case so I just want to revert to copying reshuffle and we can fix it later. I updated it to match your new reshuffle implementation, but only the one that you modified. The rest are just what they already were.

je-ik replied:
Thanks! I still wonder if we can cover the difference between `Redistribute.byKey()` and `.arbitrarily()` in tests? Or is this completely impossible? If it is impossible, why would one use the `.byKey()` variant?

kennknowles replied:
Both transforms are a hint. They are just different hints. We could run the equivalent tests for both, to make sure the runner doesn't mess one up when they directly implement. That is a good idea.

kennknowles commented on runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java:
```diff
@@ -422,6 +428,73 @@ public void translateNode(
+              getCurrentTransformName(context));
+      final Configuration partitionOptions = new Configuration();
+      partitionOptions.setString(
+          Optimizer.HINT_SHIP_STRATEGY, Optimizer.HINT_SHIP_STRATEGY_REPARTITION);
```
Just re-iterating that this is a copy-paste-modify of the Reshuffle logic. I didn't have them just share logic because I only like to do that when it is by logical necessity whereas these two transforms may at some point diverge. (but it would be fine to make them share for now)

je-ik replied:
Yes, this is actually bug in the current implementation. https://github.com/apache/beam/pull/30622

kennknowles replied:
Adjusted to match your Reshuffle implementation.

kennknowles commented on runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java:
```diff
@@ -283,6 +287,24 @@ private <K, V> void translateReshuffle(
+    DataStream<WindowedValue<KV<K, V>>> inputDataStream =
+        context.getDataStreamOrThrow(Iterables.getOnlyElement(transform.getInputsMap().values()));
+    context.addDataStream(
+        Iterables.getOnlyElement(transform.getOutputsMap().values()), inputDataStream.rebalance());
```
Would `keyBy` cause the `rebalance()` to use those keys when rebalancing? The current `Reshuffle` doesn't seem to do that.

je-ik replied:
Let's keep the streaming implementation as is.

kennknowles commented:
The new translation for `Redistribute.arbitrarily()` in the portable batch Flink runner caused a problem. I think there may be a bug in how composites are implemented because it still had transforms that consumed PCollections that had been eliminated. It is feature creep anyhow. Just translating `Redistribute.byKey()` brings it to parity. So I removed the special translation in the portable runner. Since it passed tests in the non-portable runner, I left it there.

kennknowles commented:
The problem in the portable runner was that it was not registered as a "native" transform. I don't know the details, but this influences the fuser.

kennknowles commented:
OK this is now green. The test suites that are still running were green before and have not been touched. Please review!

kennknowles commented on model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto:
```diff
@@ -813,6 +813,10 @@ message GroupIntoBatchesPayload {
   int64 max_buffering_duration_millis = 2;
 }
 
+message RedistributePayload {
```
@robertwb pinging you on this very thrilling addition to the protos

robertwb replied:
Looks fine to me.

kennknowles commented on runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java:
```diff
@@ -921,6 +926,38 @@ public void translateNode(
+      DataStream<WindowedValue<KV<K, InputT>>> inputDataSet =
+          context.getInputDataStream(context.getInput(transform));
+
+      context.setOutputDataStream(context.getOutput(transform), inputDataSet.rebalance());
```
How about we do Reshuffle and Redistribute together in a follow-up PR. This one touches so much that it is a bit sensitive to being prodded. Currently it is at parity on Flink.

je-ik replied:
Agree.

kennknowles commented on runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineTranslator.java:
```diff
@@ -917,6 +919,41 @@ private <K1, K2, V> void groupByKeyAndSortValuesHelper(
         });
 
+    registerTransformTranslator(
+        RedistributeByKey.class,
```
@robertwb you might be the person to review DataflowRunner translation? For Reshuffle we don't have a translator but a more complex rewrites to a specialized GroupByKey. I opted to _not_ do that this time but translate more directly. I added ValidatesRunner tests for Redistribute that check parity with Reshuffle, at least in terms of that test suite.

CC @scwhittle

scwhittle replied:
Should we register a RedistributeArbitrary translator as well?  At least for streaming the entirely distinct bundles with entirely unique keys is rarely a good choice.  Or a TODO if out of scope for this change.

kennknowles replied:
I do want to do a custom translation, yea. I think the best way to do it is probably a black-box Dataflow primitive rather than client side. Filed b/330519228 internally.

robertwb replied:
Is there a reason to not simply retain the existing composite (reshuffle) implementation (which can then be detected and overwritten by any runner?)

robertwb replied:
I guess I was thinking in terms of portable runners. Here we have to have a special translator to set the `PropertyNames.ALLOW_DUPLICATES` property, right?

kennknowles replied:
Yes, that's right. (and Reshuffle doesn't have a translator because it just has an override with special windowing strategy as a flag value on GBK to tell v1 to reshuffle)

kennknowles commented:
Seems that streaming wordcount is timing out at 900 seconds pretty reliably :-/

I'll keep kicking that test perhaps. Presumably it has about 1 second of compute and 15 minutes of turn up turn down though. Oof.

kennknowles commented:
Rebased passed the failure at HEAD just to get a fully green signal on this PR.

robertwb commented on runners/flink/flink_runner.gradle:
```diff
@@ -309,6 +311,8 @@ def createValidatesRunnerTask(Map m) {
 
         // Flink reshuffle override does not preserve all metadata
         excludeTestsMatching 'org.apache.beam.sdk.transforms.ReshuffleTest.testReshufflePreservesMetadata'
+        // Flink redistribute override does not preserve all metadata
```
This seems bad. Is it not possible to fix this (e.g. by reifying the metadata if needed)? I see that reshuffle didn't, but ideally the new, improved transform should do things the right way.

kennknowles replied:
OK I re-added and it succeeded. I think I added this during debugging. The non-portable runner works, if I have invoked the tests right. I'll watch the CI results.

robertwb commented on runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java:
```diff
@@ -423,6 +431,76 @@ public void translateNode(
+      final CoderTypeInformation<WindowedValue<KV<K, InputT>>> outputType =
+          ((CoderTypeInformation<WindowedValue<KV<K, InputT>>>) inputDataSet.getType())
+              .withPipelineOptions(context.getPipelineOptions());
+      // We insert a NOOP here to initialize the FileSystems via the above CoderTypeInformation.
```
Is there an upstream bug or other reference about why this hack is needed?

kennknowles replied:
TBH I'd rather separate this investigation to get the Redistribute transform working.

robertwb commented on runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java:
```diff
@@ -301,6 +305,24 @@ private <K, V> void translateReshuffle(
+    DataStream<WindowedValue<KV<K, V>>> inputDataStream =
+        context.getDataStreamOrThrow(Iterables.getOnlyElement(transform.getInputsMap().values()));
+    context.addDataStream(
+        Iterables.getOnlyElement(transform.getOutputsMap().values()), inputDataStream.rebalance());
```
Where is the key being used here? (This implementation looks identical to the one below.)

robertwb replied:
Looks identical to the code above as well. I suppose my preference is to not duplicate code unless there's a god reason for divergence or it's difficult to do so. E.g. above why not just write

```
translatorMap.put(PTransformTranslation.REDISTRIBUTE_BY_KEY_URN, this::translateReshuffle);
translatorMap.put(PTransformTranslation.REDISTRIBUTE_ARBITRARILY_URN, this::translateReshuffle);
```

which makes it clear that these are all implemented the same rather than trying to pattern match and see what the differences, if any, are.

It's not logical necessity, but saying "this (same) implementation is good enough for this usecase as well."

robertwb replied:
But if you feel strongly, no need for this to be a blocker.

kennknowles replied:
Made a long post. I very much view their current "sameness" as coincidence that can diverge. I want to freeze Reshuffle where it is and Redistribute be the place that evolves to be different. (for example incorporating more of Jan's ideas for Flink, and using new Dataflow features for Dataflow)

robertwb commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/RedistributeByKeyTranslator.java:
```diff
@@ -0,0 +1,67 @@
+    @Override
+    public boolean test(RunnerApi.PTransform pTransform) {
+      return false;
+      // Re-enable after https://github.com/apache/beam/issues/21188 is completed
```
Isn't this bug what this PR is attempting to address?

kennknowles replied:
Just removed this comment. It is simply not a native transform, which is a hack someone added to support having transforms that are not a Beam primitive.

robertwb commented on runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ReshuffleTranslator.java:
```diff
@@ -42,6 +42,16 @@
+  }
+
+  ReshuffleTranslator() {
+    this("rhfl-");
```
Was the dropping of the s in "rshfl" intentional?

kennknowles replied:
fixed

robertwb commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Redistribute.java:
```diff
@@ -0,0 +1,300 @@
+      // org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Hashing.smear().
+      int hashOfShard = 0x1b873593 * Integer.rotateLeft(shard * 0xcc9e2d51, 15);
+      if (numBuckets != null) {
+        UnsignedInteger unsignedNumBuckets = UnsignedInteger.fromIntBits(numBuckets);
```
Why not just do something like Math.abs(hashOfShard) % numBuckets?

kennknowles replied:
In other words I'm just trying to not rock the boat. I think this is update compatible in the sense that if we change the shard assignment for redistribute arbitrarily we are fine.

robertwb commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/construction/PTransformTranslation.java:
```diff
@@ -119,6 +119,9 @@ public class PTransformTranslation {
   public static final String COMBINE_PER_KEY_TRANSFORM_URN = "beam:transform:combine_per_key:v1";
   public static final String COMBINE_GLOBALLY_TRANSFORM_URN = "beam:transform:combine_globally:v1";
   public static final String RESHUFFLE_URN = "beam:transform:reshuffle:v1";
+  public static final String REDISTRIBUTE_BY_KEY_URN = "beam:transform:redistribute_by_key:v1";
```
Should these get added as constants in the proto files?

kennknowles replied:
I'll do a separate change to perhaps migrate these constants to use the constants in the proto file.

kennknowles commented:
BTW the test failures are an issue with how GHA is selecting a JVM it look like. We have precommit and postcommits passing

robertwb commented on runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingTransformTranslators.java:
```diff
@@ -162,6 +162,11 @@ class FlinkStreamingTransformTranslators {
         new CreateViewStreamingTranslator());
 
     TRANSLATORS.put(PTransformTranslation.RESHUFFLE_URN, new ReshuffleTranslatorStreaming());
+    TRANSLATORS.put(
```
Same exact comment here.

robertwb commented on runners/spark/src/main/java/org/apache/beam/runners/spark/translation/TransformTranslator.java:
```diff
@@ -788,6 +852,8 @@ public String toNativeString() {
     EVALUATORS.put(PTransformTranslation.CREATE_VIEW_TRANSFORM_URN, createPCollView());
     EVALUATORS.put(PTransformTranslation.ASSIGN_WINDOWS_TRANSFORM_URN, window());
     EVALUATORS.put(PTransformTranslation.RESHUFFLE_URN, reshuffle());
+    EVALUATORS.put(PTransformTranslation.REDISTRIBUTE_ARBITRARILY_URN, redistributeArbitrarily());
```
Same comment as for Flink (and below).

kennknowles commented:
Yea, I actually do feel strongly about the one remaining issue. I get what you are saying about directly re-using the reshuffle translator as a way of expressing "the reshuffle translation is good enough for this use case", and this is a point where I diverge with many programmers.

One helpful way I think about from the perspective of both Reshuffle and Redistribute translations being perpetual works-in-progress. If someone goes in and messes with the function "reshuffleTranslator" in either Flink or Dataflow, I don't expect them to be aware of all the places it is used. They should edit with confidence that the blast radius is just Reshuffle. So having separate-but-coincidentally-currently-identical pieces of code achieves that. Contrast that with the case where we intend one to always evolve with the other, due to logical necessity. In that case, we would also want them to only be thinking about Reshuffle, but anything implemented via Reshuffle would get the changes, because we want them to. So in this particular case, I want them to be as independent as possible. In fact, I want to actually re-deprecate Reshuffle as soon as we get RequiresStableInput implemented fully ;-) and I also want to drive all at-least-once usage to Redistribute.

Abacn commented:
Looks like several VR PostCommits failing after this PR:

https://github.com/apache/beam/actions/workflows/beam_PostCommit_Java_ValidatesRunner_ULR.yml?query=event%3Aschedule

https://github.com/apache/beam/actions/workflows/beam_PostCommit_Java_PVR_Samza.yml?query=event%3Aschedule

https://github.com/apache/beam/actions/workflows/beam_PostCommit_Java_PVR_Flink_Streaming.yml?query=event%3Aschedule

https://github.com/apache/beam/actions/workflows/beam_PostCommit_Java_PVR_Spark3_Streaming.yml?query=event%3Aschedule

https://github.com/apache/beam/actions/workflows/beam_PostCommit_Java_PVR_Spark_Batch.yml?query=event%3Aschedule



example error

https://github.com/apache/beam/actions/runs/8881963750/job/24385538170

```
> Task :runners:spark:3:job-server:validatesPortableRunnerStreaming

org.apache.beam.sdk.transforms.RedistributeTest > testRedistributeAfterSessionsAndGroupByKey FAILED
    java.lang.RuntimeException at RedistributeTest.java:265

org.apache.beam.sdk.transforms.RedistributeTest > testRedistributeAfterFixedWindowsAndGroupByKey FAILED
    java.lang.RuntimeException at RedistributeTest.java:286

org.apache.beam.sdk.transforms.RedistributeTest > testRedistributePreservesMetadata FAILED
    java.lang.RuntimeException at RedistributeTest.java:244

org.apache.beam.sdk.transforms.RedistributeTest > testRedistributeAfterSlidingWindowsAndGroupByKey FAILED
    java.lang.RuntimeException at RedistributeTest.java:307

org.apache.beam.sdk.transforms.RedistributeTest > testRedistributeAfterFixedWindows FAILED
    java.lang.RuntimeException at RedistributeTest.java:327

org.apache.beam.sdk.transforms.RedistributeTest > testRedistributeAfterSlidingWindows FAILED
    java.lang.RuntimeException at RedistributeTest.java:347
```

kennknowles commented:
Ah, I neglected the PVR suites.

kennknowles commented:
I am opening a revert in a minute but taking a look also.

 - Samza PVR is existing issue with Reshuffle so sickbay is parity
 - ULR seems to have some completely wild incorrect implementation (the Redistribute is a valid composite that ULR doesn't do anything special)
 - Flink PVR is a problem actually I fixed during the iteration on the pull request but reverted according to reviewer comments and I guess when I rebased the trigger file no longer had a delta so it seemed green
 - Spark and Spark 3 PVR is parity and needs to be sickbayed

I think I want to roll forward but without the runner translations. These are valid composites. When I first pushed it up, though, most runners failed. So there may be an issue with the composite definition.
